pretty/src/layout/boundset.rs

```diff
@@ -82,63 +112,62 @@ where
     /// Pick the best (i.e., smallest) Bound that fits within the
     /// given width. Panics if none fit.
     #[cfg(test)]
-    pub(super) fn fit_width(&self, width: Col) -> (Bound, T) {
+    pub fn fit_width(&self, width: Col) -> (Bound, T) {
         let bound = Bound::infinite_scroll(width);
         self.fit_bound(bound)
     }
 }
 
-impl<T> fmt::Debug for BoundSet<T>
-where
-    T: Clone,
-{
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let set: Vec<&Bound> = self.set.iter().map(|(bound, _)| bound).collect();
-        write!(f, "{:?}", set)
-    }
-}
-
 /// Iterator over Bounds in a BoundSet.
-pub struct BoundIter<'a, T: 'a>
+pub struct BoundSetIter<T>
 where
-    T: Clone,
+    T: Clone + fmt::Debug,
 {
-    set: &'a Vec<(Bound, T)>,
-    i: usize,
+    staircases: vec::IntoIter<Staircase<T>>,
+    bounds: Option<StaircaseIter<T>>,
 }
 
-impl<'a, T> Iterator for BoundIter<'a, T>
+impl<T> Iterator for BoundSetIter<T>
 where
-    T: Clone,
+    T: Clone + fmt::Debug,
 {
     type Item = (Bound, T);
     fn next(&mut self) -> Option<(Bound, T)> {
-        if self.i >= self.set.len() {
-            None
-        } else {
-            self.i += 1;
-            Some(self.set[self.i - 1].clone())
+        while self.bounds.is_none() {
+            match self.staircases.next() {
+                None => return None,
+                Some(staircase) => self.bounds = Some(staircase.into_iter()),
+            }
+        }
```

Comment:
This can be an `if`, right? Looks like the body is guaranteed to run at most once.

Reply:
:+1: